pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,129 @@
+limitations under the License.
+*/
+
+package authorizer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kaudit "k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/client-go/kubernetes"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpauth "github.com/kcp-dev/kcp/pkg/authorization"
+	"github.com/kcp-dev/kcp/pkg/authorization/delegated"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+type maximalPermissionAuthorizer struct {
+	getAPIExport           func(ctx context.Context, clusterName, apiExportName string) (*apisv1alpha1.APIExport, error)
+	newDelegatedAuthorizer func(clusterName string) (authorizer.Authorizer, error)
+}
+
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, kcpClusterClient kcpclient.ClusterInterface) authorizer.Authorizer {
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(ctx context.Context, clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return kcpClusterClient.Cluster(logicalcluster.New(clusterName)).ApisV1alpha1().APIExports().Get(ctx, apiExportName, v1.GetOptions{})
+		},
+		newDelegatedAuthorizer: func(clusterName string) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(logicalcluster.New(clusterName), deepSARClient)
+		},
+	}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (dec authorizer.Decision, reason string, err error) {
+	defer func() {
+		auditReason := reason
+		if err != nil {
+			auditReason = fmt.Sprintf("%v, error: %v", reason, err)
+		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionString(dec),
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, auditReason,
+		)
+	}()
+
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("invalid API domain key: %q", apiDomainKey)
+	}
+	apiExportCluster := parts[0]
+	apiExportName := parts[1]
+
+	apiExport, err := a.getAPIExport(ctx, apiExportCluster, apiExportName)
+	if err != nil && errors.IsNotFound(err) {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("API export %q not found: %q", apiDomainKey)
+	}
+
+	if apiExport.Spec.MaximalPermissionPolicy == nil {
+		return authorizer.DecisionAllow, "no maximal permission policy defined", nil
+	}
+
+	if apiExport.Spec.MaximalPermissionPolicy.Local == nil {
+		return authorizer.DecisionAllow, "no maximal local permission policy defined", nil
```

Comment:
does the reason end up in the request response? In that case, I would say we should not leak this information. Showing it in audit of course is fine (are there audit logs for virtual apiservers? cc @davidfestal)

Reply:
In fact, they do. we have https://github.com/kcp-dev/kcp/issues/1786 which asks for more transparent messages towards clients. But I'd say I'll tackle this separately so we can first reach consensus on transparency vs. information leaking separately.

Reply:
will leave messages as is

Reply:
that is, reasons won't be exposed to the user